File: TorrentParser/TorrentParser/Dictionnaries.cpp
#include "Dictionnaries.h"
#include "Parser.h"
#include "Global.h"


Dictionnaries::Dictionnaries(Parser& p_parser)
{
	this->parser = &p_parser;
	this->wordOrDef = 0;
	parseData();
	//cout << "Fin Dico" << endl;
}


Dictionnaries::~Dictionnaries()
{
}


void Dictionnaries::parseData()
{ // Parse the datas
	//cout << "Parsedata from dico" << endl;
	for (this->parser->offset = this->parser->offset; this->parser->offset < this->parser->fileSize; this->parser->offset++)
	{
		//cout << this->parser->offset << ", " << this->parser->datas[this->parser->offset] << endl;
		if (this->parser->datas[this->parser->offset] == 'e') break;
		nextDataType(this->parser->datas[this->parser->offset]);
		//cout << ", " << this->parser->datas[this->parser->offset] << endl;
	}
}



void Dictionnaries::nextDataType(char type)
{ // Determines the type of the next encountred data
	if (type == 'd')
	{ // Next data = new dictionnary
		//cout << "new dico from dico: " << this->parser->offset << ", " << this->parser->datas[this->parser->offset] << endl;
		this->parser->offset++;
		if (this->wordOrDef == 0)
		{ // It's a word
			//cout << "new word from dico: " << this->parser->offset << ", " << this->parser->datas[this->parser->offset] << endl;
			this->word.push_back(new Global(new Dictionnaries(*this->parser)));
			this->wordOrDef = 1;
		}
		else
		{ // It's a definition
			//cout << "new def from dico: " << this->parser->offset << ", " << this->parser->datas[this->parser->offset] << endl;
			this->def.push_back(new Global(new Dictionnaries(*this->parser)));
			this->wordOrDef = 0;
		}
	}

	else if (type == 'l')
	{ // Next datas = new list
		this->parser->offset++;
		//cout << "new list from dico: " << this->parser->offset << ", " << this->parser->datas[this->parser->offset] << endl;
		if (this->wordOrDef == 0)
		{ // It's a word
			//cout << "new word from dico: " << this->parser->offset << ", " << this->parser->datas[this->parser->offset] << endl;
			this->word.push_back(new Global(new Listes(*this->parser)));
			this->wordOrDef = 1;
		}
		else
		{ // It's a definition
			//cout << "new def from dico: " << this->parser->offset << ", " << this->parser->datas[this->parser->offset] << endl;
			this->def.push_back(new Global(new Listes(*this->parser)));
			this->wordOrDef = 0;
		}
	}

	else if (type <= '\x39' && type >= '\x30')
	{ // If it's a number we look for delimiters (":") to get the length of the next string
		//cout << "get length string from dico: " << type << endl;
		string s_lengthOfNextString;
		s_lengthOfNextString += type;

		this->parser->offset++;
		for (this->parser->offset = this->parser->offset; this->parser->offset < this->parser->fileSize; this->parser->offset++)
		{
			if (this->parser->datas[this->parser->offset] == ':')
			{ // Delimiters found
				this->parser->offset++;
				break;
			}
			else s_lengthOfNextString += this->parser->datas[this->parser->offset]; // Adds the current number to the end of the string
		}

		//cout << s_lengthOfNextString << endl;
		int i_lengthOfNextString = stoi(s_lengthOfNextString); // Converts the string to int
		getString(i_lengthOfNextString);
	}

	else if (type == '-' || type == 'i')
	{ // It's an integer (which constitutes a definition)
		//cout << "get an integer from dico: " << type << endl;
		string s_integer;
		this->parser->offset++;

		if (type == '-')
		{
			this->parser->offset++; // After the "-" there's always an "i" which is the begin tag of an integer
			s_integer += type;
		}

		for (this->parser->offset = this->parser->offset; this->parser->offset < this->parser->fileSize; this->parser->offset++)
		{
			if (this->parser->datas[this->parser->offset] == 'e')
			{ // Delimiters found
				break;
			}
			else s_integer += this->parser->datas[this->parser->offset]; // Adds the current number to the end of the string
		}

		//cout << s_integer << endl;

		if (this->wordOrDef == 0)
		{ // It's a word
		//	cout << "new word from dico: " << this->parser->offset << ", " << this->parser->datas[this->parser->offset] << endl;
			this->word.push_back(new Global(s_integer));
			this->wordOrDef = 1;
		}
		else
		{ // It's a definition
		//	cout << "new def from dico: " << this->parser->offset << ", " << this->parser->datas[this->parser->offset] << endl;
			this->def.push_back(new Global(s_integer));
			this->wordOrDef = 0;
		}
	}
}



void Dictionnaries::getString(int lengthOfString)
{ // Gets the string corresponding to the length

	string info;
	for (int i = 0; i < lengthOfString; i++)
	{
		info  += this->parser->datas[this->parser->offset];
		this->parser->offset++;
	}
	//cout << "string: " << info << ", length: " << lengthOfString << endl;

	this->parser->offset--; // Sets the cursor to the previous character for a proper analysis (to not interfer with the offset++ of the parseData function))

	if (this->wordOrDef == 0)
	{ // It's a word
	//	cout << "new word from dico: " << this->parser->offset << ", " << this->parser->datas[this->parser->offset] << endl;
		this->word.push_back(new Global(info));
		this->wordOrDef = 1;
	}
	else
	{ // It's a definition
	//	cout << "new def from dico: " << this->parser->offset << ", " << this->parser->datas[this->parser->offset] << endl;
		this->def.push_back(new Global(info));
		this->wordOrDef = 0;
	}
}

void Dictionnaries::printDico()
{
	cout << "[ ";
	for (unsigned int i = 0; i < word.size(); i++)
	{
		word.at(i)->printGlobal();
		cout << " => ";
		def.at(i)->printGlobal();
		cout << endl;
	}
	cout << " ]" << endl;
}

File: TorrentApp/src/main/java/TorrentApp/TorrentApp/Dictionnaries.java
package TorrentApp.TorrentApp;

import java.util.Vector;

import org.apache.log4j.Logger;


public class Dictionnaries implements Common {

	///////////////
	// Attributs //
	///////////////
	
	Parser parser; 		 // The common parser where are stored all the datas, file size and the offset
	Vector<Global> word; // A word of dictionnary. It could be a List, a string and even a dictionnary
	Vector<Global> def;  // The definition of corresponding word. It could be a List, a string and even a dictionnary
	int wordOrDef; 		 // A flag used to know if the current string is a word or a definition (0 = word, 1 = definition)
	private static final Logger LOGGER = Logger.getLogger(Dictionnaries.class);

	
	//////////////////
	// Constructeur //
	//////////////////

	public Dictionnaries(Parser p_parser)
	{
		parser = p_parser;
		word = new Vector<Global> ();
		def = new Vector<Global> ();
		wordOrDef = 0;
		parseData();
	}
	
	
	//////////////
	// Méthodes //
	//////////////

	public void parseData()
	{ // Parse the datas
			
		LOGGER.debug("Parse data from dico");

		for (parser.offset = parser.offset; parser.offset < parser.fileSize; parser.offset++)
		{
			
			LOGGER.debug(parser.offset + ", " + parser.datas[parser.offset]);

			if (parser.datas[parser.offset] == 'e') break;
			nextDataType(parser.datas[parser.offset]);
			
			LOGGER.debug(", " + parser.datas[parser.offset-1]);
		}
		
	}

	
	
	public void nextDataType(char type)
	{ // Determines the type of the next encountred data
		if (type == 'd')
		{ // Next data = new dictionnary
			
			LOGGER.debug("new dico from dico: " + parser.offset + ", " + parser.datas[parser.offset]);

			parser.offset++;
			if (wordOrDef == 0)
			{ // It's a word
				
				LOGGER.debug("new word from dico: " + parser.offset + ", " + parser.datas[parser.offset]);

				word.add(new Global(new Dictionnaries(parser)));
				wordOrDef = 1;
			}
			else
			{ // It's a definition
				
				LOGGER.debug("new def from dico: " + parser.offset + ", " + parser.datas[parser.offset]);

				def.add(new Global(new Dictionnaries(parser)));
				wordOrDef = 0;
			}
		}

		else if (type == 'l')
		{ // Next datas = new list

			parser.offset++;
			
			LOGGER.debug("new list from dico: " + parser.offset + ", " + parser.datas[parser.offset]);
			if (wordOrDef == 0)
			{ // It's a word
				
				LOGGER.debug("new word from dico: " + parser.offset + ", " + parser.datas[parser.offset]);
				word.add(new Global(new Listes(parser)));
				wordOrDef = 1;
			}
			else
			{ // It's a definition
				
				LOGGER.debug("new def from dico: " + parser.offset + ", " + parser.datas[parser.offset]);

				def.add(new Global(new Listes(parser)));
				wordOrDef = 0;
			}
		}

		else if (type <= 0x39 && type >= 0x30)
		{ // If it's a number we look for delimiters (":") to get the length of the next string
			
			LOGGER.debug("get length string from dico: " + type);

			String s_lengthOfNextString = "";
			s_lengthOfNextString += type;

			parser.offset++;
			for (parser.offset = parser.offset; parser.offset < parser.fileSize; parser.offset++)
			{
				if (parser.datas[parser.offset] == ':')
				{ // Delimiters found
					parser.offset++;
					break;
				}
				else s_lengthOfNextString += parser.datas[parser.offset]; // Adds the current number to the end of the string
			}

			
			LOGGER.debug(s_lengthOfNextString);

			int i_lengthOfNextString = Integer.parseInt(s_lengthOfNextString); // Converts the string to int
			getString(i_lengthOfNextString);
		}

		else if (type == '-' || type == 'i')
		{ // It's an integer (which constitutes a definition)
			
			LOGGER.debug("get an integer from dico: " + type);
			
			String s_integer = "";
			parser.offset++;

			if (type == '-')
			{
				parser.offset++; // After the "-" there's always an "i" which is the begin tag of an integer
				s_integer += type;
			}

			for (parser.offset = parser.offset; parser.offset < parser.fileSize; parser.offset++)
			{
				if (parser.datas[parser.offset] == 'e')
				{ // Delimiters found
					break;
				}
				else s_integer += parser.datas[parser.offset]; // Adds the current number to the end of the string
			}

			
			LOGGER.debug(s_integer);

			if (wordOrDef == 0)
			{ // It's a word
				
				LOGGER.debug("new word from dico: " + parser.offset + ", " + parser.datas[parser.offset]);
				
				word.add(new Global(s_integer));
				wordOrDef = 1;
			}
			else
			{ // It's a definition
				
				LOGGER.debug("new def from dico: " + parser.offset + ", " + parser.datas[parser.offset]);
				
				def.add(new Global(s_integer));
				wordOrDef = 0;
			}
		}
		
	}

	
	
	public void getString(int lengthOfString)
	{
		// Gets the string corresponding to the length

		String info = "";
		for (int i = 0; i < lengthOfString; i++)
		{
			info  += parser.datas[parser.offset];
			parser.offset++;
		}
		
		LOGGER.debug("string: " + info + ", length: " + lengthOfString);

		parser.offset--; // Sets the cursor to the previous character for a proper analysis (to not interfer with the offset++ of the parseData function))

		if (wordOrDef == 0)
		{ // It's a word
			
			LOGGER.debug("new word from dico: " + parser.offset + ", " + parser.datas[parser.offset]);
			
			word.add(new Global(info));
			wordOrDef = 1;
		}
		else
		{ // It's a definition
		 	
		 	LOGGER.debug("new def from dico: " + parser.offset + ", " + parser.datas[parser.offset]);
			
			def.add(new Global(info));
			wordOrDef = 0;
		}
	}

	
	
	public void printDico()
	{
		System.out.print("[ ");
		for (int i = 0; i < word.size(); i++)
		{
			word.elementAt(i).printGlobal();
			System.out.print(" => ");
			def.elementAt(i).printGlobal();
			if (i < word.size()-1) System.out.print(", ");
		}
		System.out.print(" ]");
	}
}
